fendermint/ipc/src/parent.rs

```diff
@@ -0,0 +1,202 @@
+
+impl PollingParentSyncer {
+    fn read_cache<F, T>(&self, f: F) -> T
+    where
+        F: Fn(&ParentSyncerCache) -> T,
+    {
+        let cache = self.cache.read().unwrap();
+        f(&cache)
+    }
+
+    /// Start the proof of finality listener in the background
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        match self
+            .started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => return Err(anyhow!("already started")),
+        }
+
+        let parent_syncer = self.ipc_agent_proxy.clone();
+        let config = self.config.clone();
+        let cache = self.cache.clone();
+
+        let handle =
+            tokio::spawn(async move { sync_with_parent(config, parent_syncer, cache).await });
+        self.handle = Some(handle);
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent(
+    config: Config,
+    agent_proxy: Arc<AgentProxy>,
+    lock: LockedCache,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval));
+
+    loop {
+        interval.tick().await;
+
+        let hashes = fetch_block_hashes(&config, &agent_proxy, &lock).await?;
+        update_top_down_msgs(&config, &agent_proxy, &lock).await?;
+        update_membership(&config, &agent_proxy, &lock).await?;
+
+        let mut cache = lock.write().unwrap();
+        for r in hashes {
+            cache.block_hash.insert_after_lower_bound(r.0, r.1);
+        }
+    }
+}
+
+async fn update_top_down_msgs(
+    _config: &Config,
+    _agent_proxy: &Arc<AgentProxy>,
+    _lock: &LockedCache,
+) -> anyhow::Result<Vec<CrossMsg>> {
+    Ok(vec![])
+}
+
+async fn update_membership(
+    _config: &Config,
+    _agent_proxy: &Arc<AgentProxy>,
+    _lock: &LockedCache,
+) -> anyhow::Result<ValidatorSet> {
+    Ok(ValidatorSet::default())
+}
+
+async fn fetch_block_hashes(
+    config: &Config,
+    agent_proxy: &Arc<AgentProxy>,
+    lock: &LockedCache,
+) -> anyhow::Result<Vec<(BlockHeight, Vec<u8>)>> {
+    let latest_height = match agent_proxy.get_chain_head_height().await {
+        Ok(h) => h,
+        Err(e) => {
+            tracing::warn!("cannot fetch parent chain head due to {e}");
+
+            // not throw errors, caller will retry
+            return Ok(vec![]);
+        }
+    };
+
+    if latest_height < config.chain_head_delay {
+        tracing::debug!("latest height not more than the chain head delay");
+        return Ok(vec![]);
+    }
+
+    let starting_height = {
+        let cache = lock.read().unwrap();
+        // if cache.latest_height() is None, it means we have not started fetching any heights
+        // we just use the latest height minus a lower bound as the parent view
+        let starting_height = cache.block_hash.upper_bound().unwrap_or(max(
+            1,
+            latest_height.saturating_sub(config.chain_head_lower_bound),
+        ));
```

Comment:
It's great that you thought about where to start from. I do think though that this behaviour could be improved. 

After a restart, syncing with the parent should resume from where it was left off, that is, before the syncer is started, we should look up the state of our contract that ultimately tracks the application of top-down messages and retrieve the last known final block height and nonce, and go from there. Otherwise you might stop your node for maintenance, bring it back later, and miss some of the messages simply because more than the `chain_head_lower_bound` amount of time has elapsed.

Another adverse scenario is when you stop your node, restart quickly, and then it replays messages which you already had by effectively resetting the lower bound to an earlier value!